blog/2022-05-04-announcing-stryker-6-effortless-sauce-mutation.md

```diff
@@ -0,0 +1,257 @@
+
+The `@stryker-mutator/karma-runner` and `@stryker-mutator/jest-runner` already worked with ESM (for Jest, you will need to configure the [`--experimental-vm-modules`](../docs/stryker-js/jest-runner#ecmascript-modules) flag), but our node-based runners did not. This changes with version 6.
+
+You configure it like this:
+
+```json
+
+```
+
+As you can see _no configuration change_ is needed. Updating StrykerJS is enough 😁.
+
+## ⏩ Hot reload
+
+As you might know, StrykerJS uses mutation switching (sometimes called mutant schemata), meaning Stryker inserts all mutants into your source code yet only activates them one at a time.
+
+It looks something like this:
+
+```ts
+function add(a, b) {
+  if (global.activeMutant === 0) {
+    // 👾 BlockStatement mutator
+  } else {
+    return global.activeMutant === 1
+      ? a - b // 👽 ArithmeticExpression mutator
+      : a + b;
+  }
+}
+```
+
+Next, Stryker creates _worker_ processes. A worker process is responsible for activating a mutant, running tests, and reporting the result to the main process. To do that, it uses the test runner of your choice, for example, [mocha](https://mochajs.org/). For NodeJS-based test runners, it would look something like this:
+
+(old process)
+
+1. Activate the mutant
+1. _Load all files_, using `require`
+1. Run the tests with the test runner of choice.
+1. _Unload all files_, using [`delete require.cache['path/to/cjs/module']`](https://nodejs.org/api/modules.html#requirecache) to prepare for the next run.
+1. Report the result
+
+As you can see, _all files are loaded and unloaded_ for each mutant run. This way of reloading [won't ever work with native ESM](https://github.com/nodejs/tooling/issues/51), as there is simply no `cache` we can `delete` the loaded files from 🤷‍♀️.
+
+The reloading is also expensive. Think about it: if your project contains 1k files, with a total of 10k mutants, it could mean a total of _ten million file IO actions_. That's why mutation testing with StrykerJS would slow down considerably when the size of your project grows.
+
+When you think about it, it is also **unnecessary**; why not simply keep all files loaded? Since they contain all the mutants anyway. This new process looks something like this:
+
+(new process)
+
+1. (once) Load the code using `import`
+1. Activate the mutant
+1. Run the tests with the test runner of choice.
+1. Report the result
+
+The result is a significant performance improvement 🐇. You can clearly see the difference when running StrykerJS on Stryker's core itself:
+
+| Old                                         |
+| ------------------------------------------- |
+| ![stryker-v5](/images/blogs/stryker-v5.gif) |
+
+| New \*                                      |
+| ------------------------------------------- |
+| ![stryker-v6](/images/blogs/stryker-v6.gif) |
+
+\* using `--ignoreStatic`
+
+That's a whopping 70% performance improvement 🤯.
```

Comment:
👏👏👏

Reply:
Yeah, although we had to disable the typescript checker to achieve this. The checker is so slow it takes >90% of the total time, so the result would be far less spectacular ;)